File: backend/geometry/geometry.go
package geometry

import "math"

// Point is a 2D point/vector in world space. The polygon center is (0,0).
type Point struct {
	X, Y float64
}

// Segment is a line segment defined by two points.
type Segment struct {
	A, B Point
}

// Vector helpers

func Sub(a, b Point) Point         { return Point{a.X - b.X, a.Y - b.Y} }
func Add(a, b Point) Point         { return Point{a.X + b.X, a.Y + b.Y} }
func Mul(a Point, s float64) Point { return Point{a.X * s, a.Y * s} }
func Dot(a, b Point) float64       { return a.X*b.X + a.Y*b.Y }
func Len(a Point) float64          { return math.Hypot(a.X, a.Y) }
func Norm(a Point) Point {
	l := Len(a)
	if l == 0 {
		return Point{}
	}
	return Point{a.X / l, a.Y / l}
}
func Perp(a Point) Point { return Point{-a.Y, a.X} }

// GeneratePolygon returns the vertices of a regular polygon with `sides` sides,
// centered at (0,0) with circumradius `radius`.
// The first vertex is placed at the top (-90 degrees). Vertices are returned in
// counter-clockwise order (in standard math coordinates, +Y up).
func GeneratePolygon(sides int, radius float64) []Point {
	if sides < 3 {
		sides = 3
	}
	vertices := make([]Point, 0, sides)
	for i := 0; i < sides; i++ {
		angle := -math.Pi/2 + float64(i)*2*math.Pi/float64(sides)
		vertices = append(vertices, Point{
			X: radius * math.Cos(angle),
			Y: radius * math.Sin(angle),
		})
	}
	return vertices
}

// ChamferVertices cuts off each corner of the polygon with a straight line placed
// at distance `chamferRadius` from the vertex along both adjacent edges.
// It returns an ordered list of 2*N points describing the chamfered polygon.
//
// The returned points alternate:
//   - even-indexed segments [2i, 2i+1] are the shortened original edges
//     (these become the goal faces, face i);
//   - odd-indexed segments [2i+1, 2i+2] are the chamfer cuts.
func ChamferVertices(vertices []Point, chamferRadius float64) []Point {
	n := len(vertices)
	if n < 3 {
		return vertices
	}
	// For each vertex i:
	//   a[i] lies on edge (i-1 -> i) at distance r from vertex i
	//   b[i] lies on edge (i -> i+1) at distance r from vertex i
	a := make([]Point, n)
	b := make([]Point, n)
	for i := 0; i < n; i++ {
		prev := vertices[(i-1+n)%n]
		cur := vertices[i]
		next := vertices[(i+1)%n]
		a[i] = pointAlong(cur, prev, chamferRadius)
		b[i] = pointAlong(cur, next, chamferRadius)
	}

	// Ordered counter-clockwise traversal:
	//   b[0], a[1], b[1], a[2], ..., a[n-1], b[n-1], a[0]
	out := make([]Point, 0, 2*n)
	out = append(out, b[0])
	for i := 1; i < n; i++ {
		out = append(out, a[i], b[i])
	}
	out = append(out, a[0])
	return out
}

// pointAlong returns the point on the segment from->to at distance `dist` from `from`.
func pointAlong(from, to Point, dist float64) Point {
	d := Sub(to, from)
	l := Len(d)
	if l == 0 {
		return from
	}
	return Add(from, Mul(d, dist/l))
}

// FaceMidAngle returns the world-space angle (radians) pointing from the polygon
// center to the middle of face i of a regular N-gon.
func FaceMidAngle(sides, face int) float64 {
	return -math.Pi/2 + (float64(face)+0.5)*2*math.Pi/float64(sides)
}

// ClosestPointOnSegment returns the closest point on segment s to point p.
func ClosestPointOnSegment(p Point, s Segment) Point {
	ab := Sub(s.B, s.A)
	ap := Sub(p, s.A)
	lenSq := Dot(ab, ab)
	t := 0.0
	if lenSq > 0 {
		t = Dot(ap, ab) / lenSq
		if t < 0 {
			t = 0
		} else if t > 1 {
			t = 1
		}
	}
	return Add(s.A, Mul(ab, t))
}

// DistToSegment returns the distance from point p to segment s.
func DistToSegment(p Point, s Segment) float64 {
	return Len(Sub(p, ClosestPointOnSegment(p, s)))
}

File: frontend/src/geometry.ts
// geometry.ts — mirrors the server geometry package so the client can predict
// physics locally with the exact same arena shape.

export interface Pt {
  x: number;
  y: number;
}

export interface Seg {
  a: Pt;
  b: Pt;
}

export const PI = Math.PI;

export function sub(a: Pt, b: Pt): Pt {
  return { x: a.x - b.x, y: a.y - b.y };
}
export function add(a: Pt, b: Pt): Pt {
  return { x: a.x + b.x, y: a.y + b.y };
}
export function mul(a: Pt, s: number): Pt {
  return { x: a.x * s, y: a.y * s };
}
export function dot(a: Pt, b: Pt): number {
  return a.x * b.x + a.y * b.y;
}
export function len(a: Pt): number {
  return Math.hypot(a.x, a.y);
}
export function norm(a: Pt): Pt {
  const l = len(a);
  if (l === 0) return { x: 0, y: 0 };
  return { x: a.x / l, y: a.y / l };
}
export function perp(a: Pt): Pt {
  return { x: -a.y, y: a.x };
}

/** Vertices of a regular N-gon, first vertex at top, counter-clockwise. */
export function generatePolygon(sides: number, radius: number): Pt[] {
  if (sides < 3) sides = 3;
  const out: Pt[] = [];
  for (let i = 0; i < sides; i++) {
    const angle = -PI / 2 + (i * 2 * PI) / sides;
    out.push({ x: radius * Math.cos(angle), y: radius * Math.sin(angle) });
  }
  return out;
}

/**
 * Cuts every corner with a straight line at distance `r` from the vertex.
 * Returns 2*sides ordered points; even segments are faces, odd are chamfers.
 */
export function chamferVertices(vertices: Pt[], r: number): Pt[] {
  const n = vertices.length;
  if (n < 3) return vertices.slice();

  const a: Pt[] = new Array(n);
  const b: Pt[] = new Array(n);
  for (let i = 0; i < n; i++) {
    const prev = vertices[(i - 1 + n) % n];
    const cur = vertices[i];
    const next = vertices[(i + 1) % n];
    a[i] = pointAlong(cur, prev, r);
    b[i] = pointAlong(cur, next, r);
  }

  const out: Pt[] = [b[0]];
  for (let i = 1; i < n; i++) {
    out.push(a[i], b[i]);
  }
  out.push(a[0]);
  return out;
}

function pointAlong(from: Pt, to: Pt, dist: number): Pt {
  const d = sub(to, from);
  const l = len(d);
  if (l === 0) return { x: from.x, y: from.y };
  return add(from, mul(d, dist / l));
}

/** World-space angle from the polygon center to the middle of face `face`. */
export function faceMidAngle(sides: number, face: number): number {
  return -PI / 2 + (face + 0.5) * ((2 * PI) / sides);
}

export function closestPointOnSegment(p: Pt, s: Seg): Pt {
  const ab = sub(s.b, s.a);
  const ap = sub(p, s.a);
  const lenSq = dot(ab, ab);
  let t = 0;
  if (lenSq > 0) {
    t = dot(ap, ab) / lenSq;
    if (t < 0) t = 0;
    else if (t > 1) t = 1;
  }
  return add(s.a, mul(ab, t));
}

/** Build the full wall list (2*N segments) for a chamfered arena. */
export function buildWalls(sides: number, radius: number, chamfer: number): Seg[] {
  const v = generatePolygon(sides, radius);
  const c = chamferVertices(v, chamfer);
  const walls: Seg[] = [];
  for (let i = 0; i < c.length; i++) {
    walls.push({ a: c[i], b: c[(i + 1) % c.length] });
  }
  return walls;
}
